File: course1/week3/3_car_fueling/c/main.c
/**
 *
 * C implementation to find the car fueling problem
 *
 * (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
 *
 **/

#include <stdio.h>

typedef int Type;
typedef Type* Collection;

Type minRefills( Collection S, Type N, Type reach ){
    Type ans = 0,
         T = N-1; // (T)arget index
    for( int i=0, j=i; i < T; i=j ){
        for( j=i; j+1 < N && S[j+1] - S[i] <= reach; ++j ); // increment j to farthest reachable stop
        if( j == i )
            return -1; // ran out of gas
        ans +=( j < T )? 1 : 0;
    }
    return ans;
}

int main() {
    Type target = 0,
         reach = 0,
         N = 0;
    scanf( "%d %d %d", &target, &reach, &N );
    N += 2; // +1 for the beginning stop and +1 for the target stop
    Type stops[ N ];
    stops[ 0 ] = 0;
    for( int i=1, stop=0; i+1 < N; scanf( "%d", &stop ), stops[ i++ ]=stop );
    stops[ N-1 ] = target;
    Type ans = minRefills( stops, N, reach );
    printf( "%d\n", ans );
    return 0;
}

File: course1/week3/3_car_fueling/python/main.py
# python3

##
#
# Python3 implementation to find the most valuable combination of items using a fractional knapsack
#
# (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
#
##

class Solution:
    def minRefills( self, S, N, reach ):
        ans = 0
        i = 0
        T = N-1 # (T)arget index
        while i < T:
            j = i
            while j+1 < N and S[j+1] - S[i] <= reach:
                j += 1
            if j == i:
                return -1 # ran out of gas
            ans += 1 if j < T else 0
            i = j
        return ans

if __name__ == '__main__':
    target = int( input() )
    reach = int( input() )
    N = int( input() ) + 2 # +1 for the beginning stop and +1 for the target stop
    stops = [ 0 ]
    stops.extend( map( int, input().split() ))
    stops.append( target )
    solution = Solution()
    ans = solution.minRefills( stops, N, reach )
    print( ans )
